can't find tinyxml2::tinyxml2

## Bug report

**Required Info:**

Packages (indirectly) depending on `tinyxml2` might fail, because the taget `tinyxm2::tinyxml2` is not found. 
```
02:59:47 CMake Error at CMakeLists.txt:35 (add_library):
02:59:47   Target "rosbag2_transport" links to target "tinyxml2::tinyxml2" but the
02:59:47   target was not found.  Perhaps a find_package() call is missing for an
02:59:47   IMPORTED target, or an ALIAS target is missing?
```

This happened at `rviz_common` as well as `rosbag2_transport` and is heavily depend on the include order of packages - the order in which `find_package` is called.

See the example of rviz common here: https://github.com/ros2/rviz/pull/542/commits/b327b438aefec3b5ccd6142e1ab255de19d14be4

- Operating System:
  - Windows 10
- Installation type:
  - Source, partially provoked by building with `-DENABLE_TESTING=OFF`

#### Steps to reproduce issue
<!-- Detailed instructions on how to reliably reproduce this issue http://sscce.org/
``` code that can be copy-pasted is preferred ``` -->

Revert the commit in here: https://github.com/ros2/rviz/pull/542/commits/b327b438aefec3b5ccd6142e1ab255de19d14be4 

or look at the current Windows packaging job: https://ci.ros2.org/view/packaging/job/packaging_windows/1752/console#console-section-20

Note that for the latter, the packaging job fails because `ENABLE_TESTING=OFF`, whereas the build succeeds if tests are built.

#### Expected behavior

The order in which `find_package` doesn't matter, meaning packages can be sorted alphabetically.

#### Actual behavior

The order _does_ matter! In the example of `tinyxml2`, whenever being `find_package`'d first, the compilation goes through smoothly. Being find first, also implies being found via `find_package(pluginlib)` or any other package exporting the dependency of `tinyxml2`.

--- a/tinyxml2_vendor-extras.cmake
+++ b/tinyxml2_vendor-extras.cmake
@@ -1,4 +1,4 @@
-# Copyright 2019 Open Source Robotics Foundation, Inc.
+# Copyright 2019-2020 Open Source Robotics Foundation, Inc.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -13,3 +13,34 @@
 # limitations under the License.
 
 list(INSERT CMAKE_MODULE_PATH 0 "${tinyxml2_vendor_DIR}/Modules")
+
+if(TinyXML2_FOUND AND NOT TARGET tinyxml2::tinyxml2)
+  message(STATUS "TinyXML2 was already found when tinyxml2_vendor was included, adding missing imported target tinyxml2::tinyxml2")
+  # TINYXML2_LIBRARY is composed of debug;<path\to\debug.lib>;optimized;<path\to\release.lib>
+  # we have to extract the appropriate component based on the current configuration.
+  list(LENGTH TINYXML2_LIBRARY TINYXML_LIBRARY_LIST_LENGTH)
+  if(NOT ${TINYXML_LIBRARY_LIST_LENGTH} EQUAL 4)
+    message(FATAL_ERROR "Unable to extract the library file path from ${TINYXML2_LIBRARY}")
+  endif()
+  if(CMAKE_BUILD_TYPE MATCHES DEBUG)
+    list(GET TINYXML2_LIBRARY 0 ASSERT_DEBUG)
+    if(NOT ${ASSERT_DEBUG} STREQUAL "debug")
+      message(FATAL_ERROR "could not parse debug library path from ${TINYXML2_LIBRARY}")
+    endif()
+    list(GET TINYXML2_LIBRARY 1 TINYXML2_LIBRARY_PATH)
+  else()
+    list(GET TINYXML2_LIBRARY 2 ASSERT_OPTIMIZED)
+    if(NOT ${ASSERT_OPTIMIZED} STREQUAL "optimized")
+      message(FATAL_ERROR "could not parse library path from ${TINYXML2_LIBRARY}")
+    endif()
+    list(GET TINYXML2_LIBRARY 3 TINYXML2_LIBRARY_PATH)
+  endif()
+  if(NOT EXISTS ${TINYXML2_LIBRARY_PATH})
+    message(FATAL_ERROR "library file path ${TINYXML2_LIBRARY_PATH} does not exist")
+  endif()
+
+  add_library(tinyxml2::tinyxml2 UNKNOWN IMPORTED)
+  set_property(TARGET tinyxml2::tinyxml2 PROPERTY IMPORTED_LOCATION ${TINYXML2_LIBRARY_PATH})
+  set_property(TARGET tinyxml2::tinyxml2 PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${TINYXML2_INCLUDE_DIR})
+  list(APPEND TinyXML2_TARGETS tinyxml2::tinyxml2)
+endif()

